jlm/rvsdg/region.cpp

```diff
@@ -360,47 +361,120 @@ region::NumRegions(const jlm::rvsdg::region & region) noexcept
 }
 
 std::string
-region::ToTree(const rvsdg::region & region) noexcept
+region::ToTree(const rvsdg::region & region, const util::AnnotationMap & annotationMap) noexcept
 {
-  return ToTree(region, 0);
+  std::stringstream stream;
+  ToTree(region, annotationMap, 0, stream);
+  return stream.str();
 }
 
 std::string
-region::ToTree(const rvsdg::region & region, size_t identationDepth) noexcept
+region::ToTree(const rvsdg::region & region) noexcept
+{
+  std::stringstream stream;
+  util::AnnotationMap annotationMap;
+  ToTree(region, annotationMap, 0, stream);
+  return stream.str();
+}
+
+void
+region::ToTree(
+    const rvsdg::region & region,
+    const util::AnnotationMap & annotationMap,
+    size_t identationDepth,
+    std::stringstream & stream) noexcept
 {
-  std::string subTree;
-  auto identationChar = '-';
+  static const char identationChar = '-';
+  static const char annotationSeparator = ' ';
+  static const char labelValueSeparator = ':';
+
+  auto identationString = std::string(identationDepth, identationChar);
+  auto regionString =
+      region.IsRootRegion() ? "RootRegion" : util::strfmt("Region[", region.index(), "]");
+  auto regionAnnotationString =
+      GetAnnotationString(&region, annotationMap, annotationSeparator, labelValueSeparator);
 
   // Convert current region to a string
-  if (region.IsRootRegion())
-  {
-    subTree = "RootRegion\n";
-    identationDepth += 1;
-  }
-  else if (region.node()->nsubregions() != 1)
-  {
-    auto indentationString = std::string(identationDepth, identationChar);
-    subTree += util::strfmt(indentationString, "Region[", region.index(), "]\n");
-    identationDepth += 1;
-  }
+  stream << identationString << regionString << regionAnnotationString << '\n';
+  identationDepth += 1;
 
   // Convert the region's structural nodes with their subregions to a string
-  for (const auto & node : region.nodes)
+  for (auto & node : region.nodes)
   {
     if (auto structuralNode = dynamic_cast<const rvsdg::structural_node *>(&node))
     {
-      auto identationString = std::string(identationDepth, identationChar);
       auto nodeString = structuralNode->operation().debug_string();
-      subTree += util::strfmt(identationString, nodeString, '\n');
+      auto annotationString = GetAnnotationString(
+          structuralNode,
+          annotationMap,
+          annotationSeparator,
+          labelValueSeparator);
+      stream << identationString << nodeString << annotationString << '\n';
```

Comment:
Is this line supposed to be indented one level more than the region line? It re-uses the same indentationString, but the indentation variable has been incremented, so I'm not quite sure if this is intended.

Reply:
Great catch! I seem to have accidentally moved the line for creating the indentations above the depth increase. I checked the output of the tests again to ensure it looks correct.